### PR TITLE
Add `cuda_compiler` to `zip_keys`

### PR DESCRIPTION
**Pull request by jakirkham**
Adds `cuda_compiler` to `zip_keys` alongside `cuda_compiler_version`. This is needed as part of the move to using `cuda-nvcc_{{ target_platform }}` in the future (as opposed to `nvcc_{{ target_platform }}`). Though for now this still uses `nvcc_{{ target_platform }}` in all cases (merely providing a place to add `cuda-nvcc_{{ target_platform }}` later).

xref: https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/4400

<hr>

Checklist
* [ ] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by jakirkham** [comment]
cc @bdice @leofang @adibbley @vyasr

**Comment by jakirkham** [comment]
cc @conda-forge/core

File: recipe/conda_build_config.yaml [hunk 1]
```diff
@@ -43,6 +43,10 @@ m2w64_fortran_compiler:        # [win]
 
 cuda_compiler:                 # [linux or win]
   - nvcc                       # [linux or win]
+  - nvcc                       # [linux or win]
```
Comment (isuruf) [inline recipe/conda_build_config.yaml:46]:
```suggestion
  - nvcc                       # [linux64 or win]
```

**Review by isuruf — CHANGES_REQUESTED** [review]
You need to fix the cuda migrators too

File: recipe/conda_build_config.yaml (hunk 1 above)
Comment (jakirkham) [reply recipe/conda_build_config.yaml]: Gotcha. Changed and added to the CUDA arch migrator

**Comment by jakirkham** [comment]
Good point. Those should fixed now. Please take another look and let me know if anything else is needed

File: recipe/conda_build_config.yaml [hunk 2]
```diff
@@ -41,8 +41,12 @@ m2w64_cxx_compiler:            # [win]
 m2w64_fortran_compiler:        # [win]
   - m2w64-toolchain            # [win]
 
-cuda_compiler:                 # [linux or win]
-  - nvcc                       # [linux or win]
+cuda_compiler:                 # [linux64 or win]
+  - nvcc                       # [linux64 or win]
```
Comment (isuruf) [inline recipe/conda_build_config.yaml:45]:
```suggestion
cuda_compiler:                 # [linux or win]
  - nvcc                       # [linux or win]
```

Comment (jakirkham) [reply recipe/conda_build_config.yaml:45]: This was changed in response to this suggestion above ( https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/4407#discussion_r1182972314 )

Could you please share a bit more about why we want to revert that?

Should add the CUDA arch migrator now contains this key with appropriate length

Comment (isuruf) [reply recipe/conda_build_config.yaml:45]: You are supposed to only change that for Line 46 and below.

Comment (isuruf) [reply recipe/conda_build_config.yaml:45]: The selectors used should match that of `cuda_compiler_version`

File: recipe/conda_build_config.yaml [hunk 3]
```diff
@@ -41,8 +41,12 @@ m2w64_cxx_compiler:            # [win]
...
+  - nvcc                       # [linux64 or win]
+  - nvcc                       # [linux64 or win]
+  - nvcc                       # [linux64 or win]
+  - nvcc                       # [linux64 or win]
```
[the remark this answers is not in the document]
Comment (isuruf) [reply recipe/conda_build_config.yaml:49]: Yes

File: recipe/conda_build_config.yaml (hunk 2 above)
Comment (jakirkham) [reply recipe/conda_build_config.yaml:45]: Agreed. Had another question about that below

Comment (jakirkham) [reply recipe/conda_build_config.yaml]: Sounds like we were ok with the suggestion below. So committed that

That said, if I'm still missing something here, please let me know

**Comment by isuruf** [comment]
Can you check that this works in an existing feedstock locally by passing `-e <path/to/conda-forge-pinning-feedstock/recipe/conda_build_config.yaml` to conda-smithy rerender?

**Comment by jakirkham** [comment]
Yep, submitted PR ( https://github.com/conda-forge/ucx-split-feedstock/pull/121 ) with the result of that change

Should add the CUDA arch migrator changes were needed as well

**Comment by jakirkham** [comment]
Thanks Isuru! 🙏

**Comment by jakirkham** [comment]
Tested also in PR ( https://github.com/conda-forge/arrow-cpp-feedstock/pull/1038 ) to see if `cuda_compiler` being `None` for CPU caused any issues, but it seems [that CI job is working]( https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=699844&view=logs&j=a4934dd9-cac8-50ae-1ee1-0f1db65c0032 ). So this looks good

**Comment by jakirkham** [comment]
If anything comes up, please feel free to ping me and I will take a look 🙂

**Pull request merged**